torch/csrc/jit/codegen/cuda/lower_validation.cpp

```diff
@@ -1037,6 +1037,114 @@ void validateMma(Fusion* fusion) {
   }
 }
 
+namespace {
+
+//! Returns all swizzle ops between the set of iterdomains
+//!  in `from` and `to`.
+std::vector<Expr*> getAllSwizzlesBetween(
+    std::vector<IterDomain*> from,
+    std::vector<IterDomain*> to) {
+  auto all_expr = DependencyCheck::getAllExprsBetween(
+      {from.begin(), from.end()}, {to.begin(), to.end()});
+
+  std::vector<Expr*> all_swizzles;
+
+  std::copy_if(
+      all_expr.begin(),
+      all_expr.end(),
+      std::back_inserter(all_swizzles),
+      [](Expr* expr) {
+        return expr->getExprType().has_value() &&
+            (expr->getExprType().value() == ExprType::Swizzle2D);
+      });
+
+  return all_swizzles;
+}
+
+} // namespace
+
+void validateSwizzle(Fusion* fusion) {
+  auto used_vals = fusion->usedMathVals();
+  for (auto tv : ir_utils::filterByType<TensorView>(used_vals)) {
+    if (tv->hasSwizzleOp()) {
+      // Make sure no swizzle op is inlined:
+      auto inlined_swizzles = getAllSwizzlesBetween(
+          tv->getMaybeRFactorDomain(),
+          {tv->domain()->domain().begin(),
+           tv->domain()->domain().begin() + tv->getComputeAtPosition()});
+      TORCH_INTERNAL_ASSERT(
+          inlined_swizzles.empty(), "No support for inlined swizzles");
+
+      // Make sure thread swizzling only on sharedmem
+      if (tv->getMemoryType() != MemoryType::Shared) {
+        std::vector<IterDomain*> thread_domains;
+        std::copy_if(
+            tv->domain()->domain().begin() + tv->getComputeAtPosition(),
+            tv->domain()->domain().end(),
+            std::back_inserter(thread_domains),
+            [](IterDomain* id) {
+              // Parallelized ids can participate in index swizzle
+              //  only if the tv is stored on shared mem.
+              return id->isThreadDim();
+            });
+
+        auto thread_swizzles =
+            getAllSwizzlesBetween(tv->getMaybeRFactorDomain(), thread_domains);
+        TORCH_INTERNAL_ASSERT(
+            thread_swizzles.empty(),
+            "Need sharedmem for swizzling on thread dim");
+      }
+
+      // All swizzle ops are currently assumed to
+      //  be on the right of computeAt axes.
+      // This has been ensured in previous checks.
+      // Sharing swizzled iterdomains is much more complex
+      //  and is not a near term priority.
```

Comment:
Related to the previous comment. Does this mean we can't zCurve blocks yet?

Reply:
Not yet and in a follow up I was going to add an extra attribute on the swizzle op to support zCurve on blocks as well as register access swizzle.